### PR TITLE
:sparkles: Xml parser

### LINKED ISSUE #2: 📖 Parsing XML
# 👮‍♂️ What's needed
📖 Parsing the content of a XML file and put it to the data struct

## 📋 Subtasks
- [x] Parse the content of the file to the data structure with correct data type 

# 📜 Sources
No specific sources

exemple of xml file to parse

```xml 
<document>
    <header title="Syntaxe XML">
        <author>Fornes Leo</author>
        <date>2024-01-01</date>
    </header>
    <body>
        <paragraph>This document is a simple example of the XML syntax.</paragraph>
        <paragraph>Every syntax element is displayed in this document.</paragraph>
        <section title="header 1">
            <paragraph>This is a basic paragraph with text.</paragraph>
            <paragraph>This is a paragraph with <bold>bold</bold>, <italic>italic</italic> and <code>code</code> text.</paragraph>
            <section title="header 2">
                <codeblock>
                    <paragraph>This is a code block.</paragraph>
                </codeblock>
                <list>
                    <paragraph>list item 1</paragraph>
                    <paragraph>list item 2</paragraph>
                    <paragraph>list item 3</paragraph>
                </list>
                <paragraph>This is a paragraph with a <link url="https://www.youtube.com/watch?v=dQw4w9WgXcQ&ab_channel=RickAstley">link</link>.</paragraph>
                <paragraph>This is a paragraph with an image<image url="https://cdn-images-1.medium.com/max/697/1*tsHrUKwQXG1YZX0l957ISw.png">Text to replace image</image>.</paragraph>
                <section title="">
                    <section title="header 4">
                        <paragraph>Every syntax element can be use separately or combined.</paragraph>
                        <paragraph>Think about the different possible combinations.</paragraph>
                        <paragraph>All combined syntax elements aren't displayed in this document.</paragraph>
                    </section>
                </section>
            </section>
        </section>
    </body>
</document>
```

### PR DESCRIPTION
**Pull request by landryarki**
i added function getXmlDocument
```haskell
getXmlDocument :: String -> Maybe Document
```
that parse the given string as a Xml Document and return the data structure Document (Nothing if it failed)

i also added basic parsing function

### CODE REVIEW

File: app/Main.hs [hunk 1]
```diff
@@ -7,7 +7,5 @@
 
 module Main (main) where
 
-import Lib
-
 main :: IO ()
-main = someFunc
```
Comment (axel-eck) [inline app/Main.hs:13]: Why 😆

File: src/ParserXml.hs [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+    parseWhiteSpace
+    body <- parseBody
+    parseWhiteSpace
+    parseFlag "/document"
+    return (Document header body)
+
+getXmlDocument :: String -> Maybe Document
+getXmlDocument s = case runParser parseXmlDocument s of
+    Just (doc, "") -> Just doc
+    _ -> Nothing
```
Comment (6im0n) [inline src/ParserXml.hs]: Missing a break line :)

**Review by 6im0n — CHANGES_REQUESTED** [review]
Ok for me, just make the request change

**Review by axel-eck — APPROVED** [review]
Go for me excepting the missing trailing lines, but @6im0n already took care of reporting it to you ^^

File: src/ParserData.hs [hunk 3]
```diff
@@ -51,3 +62,91 @@ instance Monad Parser where
...
+  parseChar ','
+  b <- parseInt
+  parseChar ','
+  c <- parseInt
+  parseChar ')'
+  return (a, b, c)
+
+parseString :: String -> Parser String
+parseString [] = pure []
+parseString (x:xs) = (:) <$> parseChar x <*> parseString xs
```
Comment (Marius-P1) [inline src/ParserData.hs]:
```suggestion
parseString (x:xs) = (:) <$> parseChar x <*> parseString xs

```

File: src/ParserXml.hs (hunk 2 above)
Comment (Marius-P1) [inline src/ParserXml.hs]:
```suggestion
    _ -> Nothing
    
```

**Review by Marius-P1 — CHANGES_REQUESTED** [review]
The Parsing is working perfectly but if you can just remove your warnings it will be awesome (i think you have nearly one hundred :sweat_smile: )

File: app/Main.hs (hunk 1 above)
Comment (landryarki) [reply app/Main.hs:13]: someFunc does not exist anymore, so i removed the call of it

File: src/ParserXml.hs [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
```
Comment (axel-eck) [inline src/ParserXml.hs:186]: Missing trailing line

File: src/DataStruct.hs [hunk 5]
```diff
@@ -21,11 +22,11 @@ data Document = Document {
     }
 
 data Header = Header {
-    title :: String,
-    author :: Maybe String,
-    date :: Maybe String
+    contents :: [HeaderElement]
```
Comment (axel-eck) [inline src/DataStruct.hs:25]: Why this change, it's quite more permissive, now we can have 5 titles or 3 authors and no title. Can you explain this decision?

**Review by 6im0n — APPROVED** [review]
Ok good for me, just one question for the header why made this change for the header structure ?

**Review by Marius-P1 — APPROVED** [review]
:ok:

File: src/DataStruct.hs (hunk 5 above)
Comment (landryarki) [reply src/DataStruct.hs:25]: we shouldnt have multiple same attributs as long as we don't authorize it in the parsing, as the whys and wherefores, i redirect you to @Marius-P1, i simply follow orders (its to keep an order of what attribut come before another)
